anvil/src/x11.rs

```diff
@@ -0,0 +1,342 @@
+                                            (x as i32, y as i32).into(),
+                                            output_scale,
+                                            &log,
+                                        )?;
+                                    } else {
+                                        cursor_visible = true;
+                                    }
+                                }
+
+                                #[cfg(feature = "debug")]
+                                {
+                                    use crate::drawing::draw_fps;
+
+                                    draw_fps(renderer, frame, fps_texture, output_scale as f64, fps)?;
+                                }
+
+                                Ok(())
+                            },
+                        )
+                        .map_err(Into::<SwapBuffersError>::into)
+                        .and_then(|x| x)
+                        .map_err(Into::<SwapBuffersError>::into)
+                    {
+                        Ok(()) => {
+                            // Unbind the buffer and now let the scope end to present.
+                            if let Err(err) = renderer.unbind() {
+                                error!(log, "Error while unbinding buffer: {}", err);
+                            }
+                        }
+
+                        Err(err) => {
+                            if let SwapBuffersError::ContextLost(err) = err {
+                                error!(log, "Critical Rendering Error: {}", err);
+                                state.running.store(false, Ordering::SeqCst);
+                            }
+                        }
+                    }
+                }
+
+                Err(err) => {
+                    error!(log, "Failed to allocate buffers to present to window: {}", err);
+                    state.running.store(false, Ordering::SeqCst);
+                }
+            }
+
+            #[cfg(feature = "debug")]
+            state.backend_data.fps.tick();
+            window.set_cursor_visible(cursor_visible);
+        }
+
+        // Send frame events so that client start drawing their next frame
+        state
+            .window_map
+            .borrow()
+            .send_frames(start_time.elapsed().as_millis() as u32);
+        display.borrow_mut().flush_clients(&mut state);
+
+        if event_loop
+            .dispatch(Some(Duration::from_millis(16)), &mut state)
+            .is_err()
+        {
+            state.running.store(false, Ordering::SeqCst);
+        } else {
+            display.borrow_mut().flush_clients(&mut state);
+            state.window_map.borrow_mut().refresh();
+            state.output_map.borrow_mut().refresh();
+        }
+    }
+
+    // Cleanup stuff
+    state.window_map.borrow_mut().clear();
+
+    // TODO: Figure out why the renderer is dropped later than everything else and therefore segfaults?
+    drop(renderer);
```

Comment:
This is dumb

Reply:
Doesn't crash for me if I remove the `drop()`, but I needed the following to make things work (I get a primary node here, not a render node):
```diff
diff --git a/src/backend/x11/mod.rs b/src/backend/x11/mod.rs
index 7b6c67c7d..198c7e4ab 100644
--- a/src/backend/x11/mod.rs
+++ b/src/backend/x11/mod.rs
@@ -309,9 +309,11 @@ impl X11Surface {
         )
         .map_err(AllocateBuffersError::from)?;
 
+        /*
         if get_drm_node_type_from_fd(drm_device_fd.as_raw_fd())? != DRM_NODE_RENDER {
             todo!("Attempt to get the render device by name for the DRM node that isn't a render node")
         }
+        */
 
         // Finally create a GBMDevice to manage the buffers.
         let device = gbm::Device::new(drm_device_fd.as_raw_fd()).expect("Failed to create gbm device");
```